devito/symbolics/manipulation.py

```diff
@@ -203,6 +156,74 @@ def run(expr):
     return built + rebuilt, built
 
 
+def uxreplace(expr, rule):
+    """
+    An alternative to SymPy's ``xreplace`` for when the caller can guarantee
+    that no re-evaluations are necessary or when re-evaluations should indeed
+    be avoided at all costs (e.g., to prevent SymPy from unpicking Devito
+    transformations, such as factorization).
+
+    The canonical ordering of the arguments is however guaranteed; where this is
+    not possible, a re-evaluation will be enforced.
+
+    By avoiding re-evaluations, this function is typically much quicker than
+    SymPy's xreplace.
+    """
+    return _uxreplace(expr, rule)[0]
+
+
+def _uxreplace(expr, rule):
+    """
+    Helper for uxreplace.
+    """
+    if expr in rule:
+        return rule[expr], True
+    elif rule:
+        args = []
+        changed = False
+        for a in expr.args:
+            try:
+                ax, flag = _uxreplace(a, rule)
+                args.append(ax)
+                changed |= flag
+            except AttributeError:
+                # E.g., un-sympified numbers
+                args.append(a)
+        if changed:
+            return _uxreplace_handle(expr, args), True
+    return expr, False
+
+
+@singledispatch
+def _uxreplace_handle(expr, args):
+    return expr.func(*args)
+
+
+@_uxreplace_handle.register(sympy.Add)
+def _(expr, args):
+    if all(i.is_commutative for i in args):
+        _addsort(args)  # In-place sorting
+        return expr.func(*args, evaluate=False)
+    else:
+        return expr.func(*args)
```

Comment:
Why not just `return expr`?

Reply:
by the time I get here we know the `args` have changed, so I must rebuild `expr` with the new args

Reply:
Isn't it rebuilinding everything from scratch. Wouldn't `_raw_from_args` or something be better or a `differentiable.....` to avoid reprocessing already processed args? Just wondering, this whole uxreplace looks pretty good overall.

Reply:
it's a subtletly: not all args are commutative (see `if` above) and `raw_fromargs` does **not** handle that AFAIU, so the safe way to go is re-evaluation. I **might** be wrong though

Reply:
It does in some minimal way but probably not much impact on perf
https://github.com/sympy/sympy/blob/d3ae62701d21bb07c941609d0fd0e045e55ff809/sympy/core/operations.py#L116

Reply:
Tried in a few examples (visco, tti). Didn't see any improvement basically, probably because we rarely enter that branch ( which is good). However, I ran all tests and it seems to be OK. I pushed the change, so if you feel confident about it, we can proceed this way for me